Dart/src/com/jetbrains/lang/dart/psi/impl/DartUriElementBase.java

```diff
@@ -53,6 +57,25 @@ public PsiReference getReference() {
     final String uri = ((DartUriBasedDirective)parent).getUriString();
     final int uriOffset = ((DartUriBasedDirective)parent).getUriStringOffset();
 
+    if (DartResolver.isServerDrivenResolution()) {
+      final PsiFile refPsiFile = getContainingFile();
+      final int refOffset = getTextOffset();
+      final int refLength = getTextLength();
+      final PluginNavigationRegion region = DartResolver.INSTANCE.findRegion(refPsiFile, refOffset, refLength);
+      if (region != null) {
+        final PluginNavigationTarget target = region.getTargets().get(0);
+        final String targetPath = target.getFile();
+        final VirtualFile targetVirtualFile = LocalFileSystem.getInstance().findFileByPath(targetPath);
```

Comment:
It's kind of a general rule: return references where references are expected, see if they are resolved only when asked to resolve.
For our case it is not that important to follow this rule, so it's probably ok to return EMPTY_ARRAY.
But another reason to move file search to resolve() is that we shouldn't do extra work unless we are asked for. I suggest to pass file path to DartFileReference and to search for file (and remember result) in its resolve().